tobler/util/gadm.py

```diff
@@ -0,0 +1,95 @@
+"""Utilities for fetching data from GADM."""
+import os
+import tempfile
+import time
+from warnings import warn
+
+import geopandas as gpd
+
+
+def get_gadm(code, level=0, use_fsspec=True, gpkg=True, n_retries=3):
```

Comment:
Do we want to make `tobler` a "data" access package? My sense is that'd increase the maintenance burden, tie it to external sources that may change without us controlling it, and result in functionality that is not really what the package is about. My two cents.

Reply:
thats cool, dont mean to creep the package. tldr; i thought it might be a useful way of appealing to a more international audience (if only at least in my own contrived examples) without adding a commitment to actually support the data

i agree we don't intend for tobler to become a data access package, though that was why i tucked this into util instead of making it first-class. In general, now that i think about it, i guess my implicit view of the `util` module is that its semi-private and containts convenience functions without much (any?) guarantee of long term support. Put a little differently, I wouldn't expect anyone to install a package for the functions i stash in a utility module becuase they're either too simple (e.g. get_gadm which is just string formatting) or too far removed from the core functionality of the package (e.g. segregation has a [get_network](https://pysal.org/segregation/generated/segregation.network.get_osm_network.html#segregation.network.get_osm_network) function but i dont make any promises to support osm, nor would i expect anyone to want `segregation` just for that function). Maybe that's a bad strategy and i should think more about privatizing those functions

my thought for including `get_gadm` here was that even though `tobler` is a general library for interpolation, its genesis is the harmonization of administrative data (US census data, to be specific), and unlike other packages like [pykrige](https://geostat-framework.readthedocs.io/projects/pykrige/en/stable/examples/) or [pyinterpolate](https://github.com/szymon-datalions/pyinterpolate) or [scikit-gstat](https://scikit-gstat.readthedocs.io/en/latest/userguide/kriging.html), tobler's focus on human geography and administrative data are part of its DNA, and i think it should remain that way. What i'd like to move away from is currently all `tobler`s existing use-cases and examples are US-centric and i'd like to make it easier to use the library everywhere. E.g., we built the raster interface specifically to consume NLCD data (cause thats what our grant needed) but always intended that it be general enough for use anywhere. It took me a long time to find the Copernicus LULC data but now we can both test raster workflows and show examples of the library using global data. With get_gadm i thought it might be an easy way to demonstrate toblers value to a global audience without actually committing to providing data. (my sense is gadm is pretty solid, but if it goes away so does the function--take it up with them!)

either way i'll split it out of this PR

Reply:
i think i would make the same argument for h3fy which is a util function for the same reason. `tobler` isnt a computational geometry library either but we include the function in util because its really handy to have that capability around in tobler-based work. If h3py changed their api or took their C library private, it would increase our maintenance burden and tie us to external support in the capacity as it would if gadm made changes or went down (which is to say none. We'd just drop the function)

Reply:
I see your point. I see the value of adding little helpers like this on `utils`, I'm just concious that once something is shipped and released, there's a bit of implicit burden and expectation. While you may not want users to come to `tobler` for certain things, once it's out, the beauty of open source is folks will view it in different ways. But again, that's my personal view.

> my thought for including get_gadm here was that even though tobler is a general library for interpolation, its genesis is the harmonization of administrative data (US census data, to be specific), and unlike other packages like pykrige or pyinterpolate or scikit-gstat, tobler's focus on human geography and administrative data are part of its DNA, and i think it should remain that way. 

Totally agree. In my mind `tobler` has a clear distinctive USP in providing interpolation for _areal data_. But, the same way `gstat` doesn't have to support access to specific point datasets, I don't see it as core functionality

> What i'd like to move away from is currently all toblers existing use-cases and examples are US-centric and i'd like to make it easier to use the library everywhere. 

Again, I'm totally +1 on this. My view (which again is just a personal opinion, so it can be taken as just that) is that the way to achieve that is to integrate as closely as possible with the _data structures_ of the ecosystem, rather than tying to specific _file formats_ or data packages. When it comes to rasters, my sense (and that's why we went for it on the raster bridge in `libpysal`) is that the community consensus is centerin around `rio/xarray`, the same way it is building around `GeoDataFrame` objects for vectors. So if we build functionality fully on top of this, our users only need to be able to express whichever data they have on the standard data structure of the broader community, and they should be good to go. On the examples, that sounds also great, and this is something maybe @martinfleis and I can help with our current project, but this is more about how documentation and pedagogic tutorials are built. When it comes to core functionality of the library, I'd try to keep it as abstract/general as possible.

> i think i would make the same argument for h3fy which is a util function for the same reason. tobler isnt a computational geometry library either but we include the function in util because its really handy to have that capability around in tobler-based work. If h3py changed their api or took their C library private, it would increase our maintenance burden and tie us to external support in the capacity as it would if gadm made changes or went down (which is to say none. We'd just drop the function)

On `h3fy`, yes and no. Unlike GADM, h3 is a library based on an open indexing algorithm, not a data service. Of course if they removed functionality or make the algo proprietary things change, but in the same way I suppose that if `pandas` decides to do away with `DataFrame` objects because the Arrow project has a new shiny thing that's different. I see it more as code than as data.

Once installed, h3 does not require internet connection to be usable, and what `h3fy` does is hooking up an existing spatial indexing algorithm to `tobler`'s core USP. That's why I see it a bit more attractive than the GADM. I'm sure the GADM team does a great job at staying stable over time, but I think I've heard enough horror stories from @ljwolf on trying to maintain `cenpy`, which is a package for an official, tax-payer funded API, and still causes plenty of headaches as far as I can tell.

Reply:
ok im convinced. "strong opinions weakly held" and all that :P

> While you may not want users to come to tobler for certain things, once it's out, the beauty of open source is folks will view it in different ways. But again, that's my personal view.

thanks for the perspective. exchanges like this are why im grateful for getting to engage with more experienced devs like you.  At the moment I guess i'm more liberal about including one-off functionality because of the other side of the open-source coin--i.e. I'm still pretty comfortable saying "feature x is not core to the library and its no longer supported, but you're welcome to join the dev team and support it or fork the package if you want". Though in fairnesss, none of the packages i lead are that wildly popular and i havent had to endure the wrath of the internet hoards yet